[INTEL MKL] MKL-DNN v1.0 integration with pooling ops

nammbash opened the pull request:
MKL-DNNL v1.0 integration with pooling ops (avg & max pooling, both fwd and bwd).

nammbash committed:
Support mkldnn_v1.0 for pooling ops

nammbash commented:
@penpornk Not sure if you are reviewing other DNN 1.0 PRs if yes can you also make some comments on these please or pull this in

penpornk commented on tensorflow/core/kernels/mkl_pooling_ops_common.cc:
```diff
@@ -15,86 +15,101 @@ limitations under the License.
     context_.fwd.reset(new pooling_forward(*context_.fwd_pd, *context_.src_mem,
                                            *context_.dst_mem,
                                            *context_.ws_mem));
+#endif
```
Nit: The #if is quite long. Let's properly comment its end.
```suggestion
#endif  // ENABLE_MKLDNN_V1
```

penpornk commented on tensorflow/core/kernels/mkl_pooling_ops_common.cc:
```diff
@@ -15,86 +15,101 @@ limitations under the License.
+  context_.dst_mem.reset(new MEMORY_CONSTRUCTOR(
+      context_.fwd_pd.get()->PRIMITIVE_DESC_DST, cpu_engine_, DummyData));
+
+  // For max pooling, need to return workspace(ws) for backward computing.
```
Nit:
```suggestion
  // For max pooling, need to return workspace (ws) for backward computing.
```

penpornk commented on tensorflow/core/kernels/mkl_pooling_ops_common.cc:
```diff
@@ -106,20 +121,29 @@ void MklPoolingFwdPrimitive<T>::Execute(const T* src_data, T* dst_data,
     context_.ws_mem->set_data_handle(ws_data);
   }
+
+#ifdef ENABLE_MKLDNN_V1
```
Seems like this (lines 132-138) is used in many ops. Maybe refactor it into `mkl_util` or something? Please also use this for `bwd_primitives`.

penpornk replied:
This hasn't been addressed yet. (If it's not worth doing, please explain why. Otherwise, please add a TODO or make modifications.)

nammbash replied:
@penpornk This is what I was looking into if at all it is possible to do this in a simple way. I think you might have misunderstood mkl op context_ to be a generic class as such as the Eigen OpKernelContext. unfortunately, the mkl op context_ is specific to each op and specific to fwd and bwd prop within each of those op. Going through the MklPrimitive also did not seem like a simple fix and needs a decent redesign. Opened an internal ticket for this to be redesigned for east of use in future. So it is worth doing, but Not easily doable without a significant redesign.

penpornk replied:
But the primitives and streams in each context are of the same type, right? In here (and `mkl_conv_grad_filter_ops.cc` and `mkl_softmax_op.cc`) `fwd_primitives` `bwd_primitives` are always of type `mkldnn:primitive` and `fwd_stream` is of type `mkldnn::stream`. I'm thinking of something along the line of 
```c++
// In mkl_utils.h or some other common file.
#ifdef ENABLE_MKLDNN_V1
void execute_primitives(std::vector<mkldnn::primitive>& primitives, 
                        std::shared_ptr<stream> stream, 
                        std::vector<std::unordered_map<int, memory>>& net_args) {
  DCHECK_EQ(primitives.size(), net_args.size();
  for (size_t i = 0; i < primitives.size(); ++i) {
    primitives.at(i).execute(*stream, net_args.at(i));
  }
}
#endif  // ENABLE_MKLDNN_V1

// In this file.
#ifdef ENABLE_MKLDNN_V1
  execute_primitives(context_.fwd_primitives, context_.fwd_stream, context_.net_args);
#else 
  context_.fwd_stream->submit(context_.fwd_primitives);
#endif  // ENABLE_MKLDNN_V1
```

I think this is trivial enough to do in this PR. But if you strongly prefer doing it in a later PR, I'm okay with that too.

penpornk commented on tensorflow/core/kernels/mkl_pooling_ops_common.cc:
```diff
@@ -132,64 +156,93 @@ template class MklPoolingFwdPrimitive<bfloat16>;
     context_.bwd.reset(
         new pooling_backward(*context_.bwd_pd, *context_.diff_dst_mem,
                              *context_.ws_mem, *context_.diff_src_mem));
+#endif
```
Nit:
```suggestion
#endif  // ENABLE_MKLDNN_V1
```

penpornk reviewed (changes requested):
Thank you for the PR! I have minor comments.

penpornk commented on tensorflow/core/kernels/mkl_pooling_ops_common.cc:
```diff
@@ -132,64 +156,93 @@ template class MklPoolingFwdPrimitive<bfloat16>;
+
+#ifdef ENABLE_MKLDNN_V1
+  // Store expected primitive format.
+  context_.diff_src_fmt = MEMORY_FORMAT::any;
```
Are lines 190-191 necessary? They are already set in the constructor of the struct.

penpornk commented on tensorflow/core/kernels/mkl_pooling_ops_common.cc:
```diff
@@ -15,86 +15,101 @@ limitations under the License.
-  if (fwdParams.alg_kind == pooling_max &&
+#ifdef ENABLE_MKLDNN_V1
+  // Store expected primitive format.
+  context_.dst_fmt = MEMORY_FORMAT::any;
```
Is this necessary? This is the default value set in the struct's constructor.

penpornk replied:
This hasn't been addressed yet.

penpornk reviewed (changes requested):
I forgot to add two comments.

nammbash committed:
Review Comments 1

nammbash commented on tensorflow/core/kernels/mkl_avgpooling_op.cc:
```diff
@@ -222,45 +223,50 @@ class MklAvgPoolingGradOp : public MklPoolingBackwardOpBase<T> {
-            {{diff_dst_dims}, MklDnnType<T>(), pooling_bwd->GetDiffDstFormat()},
-            cpu_engine_);
+
+      // Check whether we need to reorder diff_dst.
```
In the interest of pushing dnn 1.0 PR's I have added a ToDo to fix this soon. I will take up the AR to  modify it on later PRs.

nammbash committed:
Merge branch 'master' into niroop/mkldnn_v1.0_pooling_ops

penpornk commented on tensorflow/core/kernels/mkl_pooling_ops_common.cc:
```diff
@@ -100,7 +100,7 @@ void MklPoolingFwdPrimitive<T>::Setup(const MklPoolingParams& fwdParams) {
                                            *context_.dst_mem,
                                            *context_.ws_mem));
-#endif
+#endif // ENABLE_MKLDNN_V1
```
Nit: (two spaces between `#endif` and `//`.
```suggestion
#endif  // ENABLE_MKLDNN_V1
```

Same for the other two `endif`s.

penpornk commented on tensorflow/core/kernels/mkl_avgpooling_op.cc:
```diff
@@ -243,6 +243,8 @@ class MklAvgPoolingGradOp : public MklPoolingBackwardOpBase<T> {
               : memory::desc(diff_dst_dims, MklDnnType<T>(),
                              this->data_format_mkldnn_);
 
+      // @TODO nammbash:Refactor (lines 249-262) common code for 
```
Nit:
```suggestion
      // TODO(nammbash): Refactor (lines 249-262) common code for 
```

penpornk reviewed (changes requested):
Thank you for the fixes! A few comments haven't been addressed yet and the `#endif` comment fixes still have wrong formatting. Sorry for the nits, I just think we still have time to get them right.

nammbash commented:
@penpornk Many has not been addressed. I am working on it. pushed a check in. More soon. Next commit will have most of the changes
Thank you!

penpornk commented:
@nammbash Sorry about that! I'll wait until you say it's ready next time.

nammbash committed:
reviews comments 2

nammbash committed:
reviews comments 3

nammbash commented:
@penpornk Thank you and sorry for pushing commits without fixing everything. I addressed all comments now. please let me know.
@rthadur if you need to start any of the CI , please do so.

penpornk commented on tensorflow/core/kernels/mkl_pooling_ops_common.cc:
```diff
@@ -228,7 +226,7 @@ void MklPoolingBwdPrimitive<T>::Setup(const MklPoolingParams& bwdParams) {
         new pooling_backward(*context_.bwd_pd, *context_.diff_dst_mem,
                              *context_.ws_mem, *context_.diff_src_mem));
-#endif //ENABLE_MKLDNN_V1
+#endif // ENABLE_MKLDNN_V1
```
Unfortunately this still isn't fixed yet. In my last review, I was trying to say that there should be two spaces between `#endif` and `//`. Since you'll have to make one more edit anyway, would you mind also fixing this? Thank you!
```suggestion
#endif  // ENABLE_MKLDNN_V1
```

nammbash replied:
and I almost made all others to single space including
```#endif  // INTEL_MKL
#endif  // TENSORFLOW_CORE_UTIL_MKL_UTIL_H_```

Luckily I did not. Fixed with two spaces.

penpornk commented on tensorflow/core/kernels/mkl_pooling_ops_common.cc:
```diff
@@ -106,20 +118,30 @@ void MklPoolingFwdPrimitive<T>::Execute(const T* src_data, T* dst_data,
   }
+
+#ifdef ENABLE_MKLDNN_V1
+  //ExecuteFwdPrimitives(context_);
```
Please delete this line.

nammbash replied:
Done! Thank you for catching it!

nammbash replied:
The suggested code changes seems reasonable and minimal changes I have changed with suggested code change. Looks clean and thank you for the suggestion.

penpornk reviewed (changes requested):
No worries about making many commits! That part is on me for not noticing you were not done yet (I got a review request so I went straight to reviewing).

nammbash committed:
reviews comments 4

nammbash committed:
reviews comments 5

nammbash committed:
reviews comments 6: Refactor common code

nammbash committed:
reviews comments 7: Refactor common code bwd + clang formatting

nammbash commented:
@penpornk Can you let me know if there are any more changes to it. Other PR's are waiting on this apprently from your comments

penpornk reviewed (approved):
Sorry for the delay. I had been in back-to-back meetings since noon. Thank you so much for the changes!

nammbash commented:
@penpornk Thank you! Will update the team internally for the changes to common files. Thanks for the review again!

penpornk commented:
@nammbash That would be very helpful. Thank you! I'll try to get this PR merged by today (if there is no test failure) and hopefully merge other PRs tomorrow.

nammbash commented:
@penpornk thank you. There is a GPU failure? unrelated to my PR!?

penpornk commented:
@nammbash Don't worry about `Windows Bazel GPU`. It's unrelated. Only `Ubuntu CPU` and `Ubuntu Sanity` need to pass. It's fine if the failed ones are all unrelated.

nammbash commented:
@penpornk @rthadur All passed except one which is unrelated to my PR. Please merge it. :) thank you.

penpornk commented:
@nammbash I'm getting `clang-format` errors. Would you mind running `clang-format` on `mkl_util.h`? This is the diff I got:
```
--- //tensorflow/core/util/mkl_util.h	(before formatting)
+++ //tensorflow/core/util/mkl_util.h	(after formatting)
@@ -728,9 +728,9 @@
     }
     return Status::OK();
   } catch (mkldnn::error& e) {
-    string error_msg = "Status: " + std::to_string(e.status) + ", message: " +
-                       string(e.message) + ", in file " + string(__FILE__) +
-                       ":" + std::to_string(__LINE__);
+    string error_msg = "Status: " + std::to_string(e.status) +
+                       ", message: " + string(e.message) + ", in file " +
+                       string(__FILE__) + ":" + std::to_string(__LINE__);
     LOG(FATAL) << "Operation received an exception: " << error_msg;
   }
 }
@@ -1250,8 +1250,8 @@
   } catch (mkldnn::error& e) {
     return Status(error::Code::INTERNAL,
                   tensorflow::strings::StrCat(
-                      "Failed to create blocked memory descriptor.", "Status: ",
-                      e.status, ", message: ", e.message));
+                      "Failed to create blocked memory descriptor.",
+                      "Status: ", e.status, ", message: ", e.message));
   }
 #else
   // We have to construct memory descriptor in a C style. This is not at all
```

nammbash commented:
@penpornk Really, I ran it for all the files with my last commit. Let me run it again and check.

nammbash commented:
@penpornk Not seeing it from my side.
Command used on my side:
> (fasterrcnnfpn) nammbash@aipg-ra-skx-193:/localdisk/niroop/tensorflow$ /opt/tensorflow/google-code-compliance/clang+llvm-3.9.0-x86_64-fedora23/bin/clang-format -style=Google tensorflow/core/util/mkl_util.h > ../temp_mkl_util.h
> (fasterrcnnfpn) nammbash@aipg-ra-skx-193:/localdisk/niroop/tensorflow$ diff tensorflow/core/util/mkl_util.h ../temp_mkl_util.h
> (fasterrcnnfpn) nammbash@aipg-ra-skx-193:/localdisk/niroop/tensorflow$

penpornk commented:
@nammbash Would you mind formatting it manually for now? Otherwise, I can't merge. (If I make these formatting changes myself, I'll need someone else to review my changes internally, which will not happen tonight.)

nammbash commented:
@penpornk I think I know what is happening. Please do a git pull of your local branch with my Latest PR changes. You are probably looking at on older version with your code you used to check or have a conflict that git pull did not replace the file.

penpornk commented on tensorflow/core/util/mkl_util.h:
```diff
@@ -728,9 +728,9 @@ inline Status ConvertMklToTF(OpKernelContext* context,
-                       string(__FILE__) + ":" + std::to_string(__LINE__);
+    string error_msg = "Status: " + std::to_string(e.status) + ", message: " +
+                       string(e.message) + ", in file " + string(__FILE__) +
+                       ":" + std::to_string(__LINE__);
```
I think the formatting change here need to be reverted.

penpornk commented on tensorflow/core/util/mkl_util.h:
```diff
@@ -1250,8 +1250,8 @@ inline Status CreateBlockedMemDescHelper(const memory::dims& dim,
-                      "Failed to create blocked memory descriptor.",
-                      "Status: ", e.status, ", message: ", e.message));
+                      "Failed to create blocked memory descriptor.", "Status: ",
+                      e.status, ", message: ", e.message));
```
This too.

penpornk reviewed (changes requested):
The whole PR was pulled less than an hour ago, I'm pretty sure we have the latest changes. Looking at the diff manually, it seems the latest formatting changes in `mkl_util.h` shouldn't have happened.

nammbash commented:
@penpornk looks like your clang is undoing my clang. 
I think we need to check both these versions. Anyways, let me try and undo the local changes I did for that file.

nammbash committed:
review comments 8: clang formatting undoing for mkl_util unrelated sections

nammbash commented:
@penpornk  done

penpornk reviewed (approved):
Thank you! Let's try then.

penpornk commented:
@nammbash Will email you the version.

penpornk commented on tensorflow/core/util/mkl_util.h:
```diff
@@ -730,7 +730,7 @@ inline Status ConvertMklToTF(OpKernelContext* context,
     string error_msg = "Status: " + std::to_string(e.status) +
                        ", message: " + string(e.message) + ", in file " +
-                       string(__FILE__) + ":" + std::to_string(__LINE__);
+                       string(__FILE__) + ":" + std::to_string(__LINE__);    
```
Still got clang-format errors. This time because of the trailing spaces. (Sorry, I didn't notice it earlier.)
```suggestion
                       string(__FILE__) + ":" + std::to_string(__LINE__);
```

mdfaijul commented on tensorflow/core/util/mkl_util.h:
```diff
@@ -2155,6 +2155,17 @@ inline bool IsConv1x1StrideNot1(memory::dims filter_dims,
+void execute_primitives(
+    std::vector<mkldnn::primitive>& primitives, std::shared_ptr<stream> stream,
+    std::vector<std::unordered_map<int, memory>>& net_args) {
+  DCHECK_EQ(primitives.size(), net_args.size();
```
@nammbash line 2162 does have a closing parenthesis! It will not build.

penpornk replied:
Thanks for the heads up! I'll fix this, along with the remaining formatting error myself then.

nammbash replied:
@penpornk thank you let me know if you still need me to do something. There is a version mismatch clearly..which I will resolve internally.
